[BUG] Random freezes during printing over serial connection (skr 1.4 turbo)

### Bug Description

I'm experiencing radom freezes when printing with Octoprint over serial connection.
Using SKR 1.4 turbo, RPI4, directly connected to TFT serial interface (SERIAL_PORT 0).
Running bugfix-2x 004bed8a7fc3ff9feb73a0ea9794635b50073c27
I am not able to identify any specific circumstances of the error - sometimes it happens on first layer, and sometimes after few hours of printing the same gcode.
Print head stops, screen (ulti controler) stops refreshing, communication freezes (there is "timeout" message in octoprint terminal window).

I'm suspecting it's related to recent serial changes, because:
- I'm not experiencing this issue when printing from SD card
- I've reverted to 811bb7997cf34f6d9559c5f281d209445c4fc123 and so far it seems to be working fine.


### Configuration Files

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5936904/Configuration.zip)

Marlin\src\MarlinCore.cpp:873:12: error: invalid conversion from 'void (*)()' to 'void*' [-fpermissive]
  873 | void * a = &Invalid<decltype(MYSERIAL1)>::nothing;
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |            |
      |            void (*)()
Marlin\src\MarlinCore.cpp: In instantiation of 'static void Invalid<T>::nothing() [with 
T = MarlinSerial]':
Marlin\src\MarlinCore.cpp:873:43:   required from here
Marlin\src\MarlinCore.cpp:872:93: error: size '4294768624' of array exceeds maximum object size '2147483647'
  872 | template <typename T> struct Invalid { static void nothing() { const char invalid[sizeof(T) - 200000]; } };
      |
  ~~~~~~~~~~^~~~~~~~
Marlin\src\MarlinCore.cpp:872:75: error: uninitialized const 'invalid' [-fpermissive]   
  872 | template <typename T> struct Invalid { static void nothing() { const char invalid[sizeof(T) - 200000]; } };
      |                                                                           ^~~~~~~
Marlin\src\MarlinCore.cpp:872:75: warning: unused variable 'invalid' [-Wunused-variable]*** [.pio\build\LPC1769\src\src\MarlinCore.cpp.o] Error 1

--- a/Marlin/src/HAL/ESP32/FlushableHardwareSerial.h
+++ b/Marlin/src/HAL/ESP32/FlushableHardwareSerial.h
@@ -22,6 +22,8 @@
 #pragma once
 
 #include <HardwareSerial.h>
+
+#include "../shared/Marduino.h"
 #include "../../core/serial_hook.h"
 
 class FlushableHardwareSerial : public HardwareSerial {

--- a/Marlin/src/HAL/ESP32/FlushableHardwareSerial.h
+++ b/Marlin/src/HAL/ESP32/FlushableHardwareSerial.h
@@ -21,8 +21,7 @@
  */
 #pragma once
 
-#ifdef ARDUINO_ARCH_ESP32
-
+#include "../shared/serial_hook.h"
 #include <HardwareSerial.h>
 #include "../../core/serial_hook.h"
 
@@ -32,5 +31,3 @@ public:
 };
 
 extern Serial0Type<FlushableHardwareSerial> flushableSerial;
-
-#endif // ARDUINO_ARCH_ESP32

--- a/Marlin/src/HAL/ESP32/FlushableHardwareSerial.h
+++ b/Marlin/src/HAL/ESP32/FlushableHardwareSerial.h
@@ -21,7 +21,6 @@
  */
 #pragma once
 
-#include "../shared/serial_hook.h"
 #include <HardwareSerial.h>
 #include "../../core/serial_hook.h"
 

--- a/Marlin/src/core/serial.h
+++ b/Marlin/src/core/serial.h
@@ -62,7 +62,7 @@ extern uint8_t marlin_debug_flags;
 //
 // Serial redirection
 //
-#define SERIAL_ALL 0x7F
+#define SERIAL_ALL 0xFF
 #if HAS_MULTI_SERIAL
   #define _PORT_REDIRECT(n,p)   REMEMBER(n,multiSerial.portMask,p)
   #define _PORT_RESTORE(n,p)    RESTORE(n)

--- a/Marlin/src/core/serial_hook.h
+++ b/Marlin/src/core/serial_hook.h
@@ -165,12 +165,12 @@ struct RuntimeSerial : public SerialBase< RuntimeSerial<SerialT> >, public Seria
   RuntimeSerial(const bool e, Args... args) : BaseClassT(e), SerialT(args...), writeHook(0), eofHook(0), userPointer(0) {}
 };
 
-// A class that's duplicating its output conditionally to 2 serial interface
+// A class that duplicates its output conditionally to 2 serial interfaces
 template <class Serial0T, class Serial1T, const uint8_t offset = 0, const uint8_t step = 1>
 struct MultiSerial : public SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > {
   typedef SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > BaseClassT;
 
-  uint8_t    portMask;
+  uint8_t    portMask = SERIAL_ALL;
   Serial0T & serial0;
   Serial1T & serial1;
 

--- a/Marlin/src/core/serial_hook.h
+++ b/Marlin/src/core/serial_hook.h
@@ -170,7 +170,7 @@ template <class Serial0T, class Serial1T, const uint8_t offset = 0, const uint8_
 struct MultiSerial : public SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > {
   typedef SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > BaseClassT;
 
-  uint8_t    portMask = 0xFF;
+  uint8_t    portMask;
   Serial0T & serial0;
   Serial1T & serial1;
 

--- a/Marlin/src/core/serial_hook.h
+++ b/Marlin/src/core/serial_hook.h
@@ -170,7 +170,7 @@ template <class Serial0T, class Serial1T, const uint8_t offset = 0, const uint8_
 struct MultiSerial : public SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > {
   typedef SerialBase< MultiSerial<Serial0T, Serial1T, offset, step> > BaseClassT;
 
-  uint8_t    portMask = SERIAL_ALL;
+  uint8_t    portMask = 0xFF;
   Serial0T & serial0;
   Serial1T & serial1;
 

--- a/Marlin/src/gcode/gcode.cpp
+++ b/Marlin/src/gcode/gcode.cpp
@@ -1062,6 +1062,7 @@ void GcodeSuite::process_subcommands_now(char * gcode) {
     static millis_t next_busy_signal_ms = 0;
     if (!autoreport_paused && host_keepalive_interval && busy_state != NOT_BUSY) {
       if (PENDING(ms, next_busy_signal_ms)) return;
+      PORT_REDIRECT(SERIAL_ALL);
       switch (busy_state) {
         case IN_HANDLER:
         case IN_PROCESS:

--- a/Marlin/src/gcode/queue.cpp
+++ b/Marlin/src/gcode/queue.cpp
@@ -272,7 +272,7 @@ void GCodeQueue::flush_and_request_resend(const serial_index_t serial_ind) {
   SERIAL_ECHOLN(serial_state[serial_ind].last_N + 1);
 }
 
-inline bool serial_data_available(int8_t index) {
+inline bool serial_data_available(uint8_t index) {
   const int a = SERIAL_IMPL.available(index);
   #if BOTH(RX_BUFFER_MONITOR, RX_BUFFER_SIZE)
     if (a > RX_BUFFER_SIZE - 2) {

--- a/Marlin/src/gcode/queue.cpp
+++ b/Marlin/src/gcode/queue.cpp
@@ -272,21 +272,7 @@ void GCodeQueue::flush_and_request_resend(const serial_index_t serial_ind) {
   SERIAL_ECHOLN(serial_state[serial_ind].last_N + 1);
 }
 
-// Multiserial already handle the dispatch to/from multiple port by itself
-inline bool serial_data_available(uint8_t index = SERIAL_ALL) {
-  if (index == SERIAL_ALL) {
-    for (index = 0; index < NUM_SERIAL; index++) {
-      const int a = SERIAL_IMPL.available(index);
-      #if BOTH(RX_BUFFER_MONITOR, RX_BUFFER_SIZE)
-        if (a > RX_BUFFER_SIZE - 2) {
-          PORT_REDIRECT(SERIAL_PORTMASK(index));
-          SERIAL_ERROR_MSG("RX BUF overflow, increase RX_BUFFER_SIZE: ", a);
-        }
-      #endif
-      if (a > 0) return true;
-    }
-    return false;
-  }
+inline bool serial_data_available(int8_t index) {
   const int a = SERIAL_IMPL.available(index);
   #if BOTH(RX_BUFFER_MONITOR, RX_BUFFER_SIZE)
     if (a > RX_BUFFER_SIZE - 2) {
@@ -294,8 +280,14 @@ inline bool serial_data_available(uint8_t index = SERIAL_ALL) {
       SERIAL_ERROR_MSG("RX BUF overflow, increase RX_BUFFER_SIZE: ", a);
     }
   #endif
-
   return a > 0;
+}
+
+// Multiserial already handles dispatch to/from multiple ports
+inline bool any_serial_data_available() {
+  LOOP_L_N(p, NUM_SERIAL)
+    if (serial_data_available(p))
+      return true;
 }
 
 inline int read_serial(const uint8_t index) { return SERIAL_IMPL.read(index); }
@@ -409,7 +401,7 @@ void GCodeQueue::get_serial_commands() {
   // send "wait" to indicate Marlin is still waiting.
   #if NO_TIMEOUTS > 0
     const millis_t ms = millis();
-    if (ring_buffer.empty() && !serial_data_available() && ELAPSED(ms, last_command_time + NO_TIMEOUTS)) {
+    if (ring_buffer.empty() && !any_serial_data_available() && ELAPSED(ms, last_command_time + NO_TIMEOUTS)) {
       SERIAL_ECHOLNPGM(STR_WAIT);
       last_command_time = ms;
     }

--- a/platformio.ini
+++ b/platformio.ini
@@ -423,7 +423,7 @@ MORGAN_SCARA            = src_filter=+<src/gcode/scara>
 HAS_MICROSTEPS          = src_filter=+<src/gcode/control/M350_M351.cpp>
 (ESP3D_)?WIFISUPPORT    = AsyncTCP, ESP Async WebServer
   ESP3DLib=https://github.com/luc-github/ESP3DLib.git
-  arduinoWebSockets=https://github.com/Links2004/arduinoWebSockets/archive/2.3.4.zip
+  arduinoWebSockets=links2004/WebSockets@2.3.4
   ESP32SSDP=https://github.com/luc-github/ESP32SSDP.git
   lib_ignore=ESPAsyncTCP
 

--- a/platformio.ini
+++ b/platformio.ini
@@ -422,9 +422,9 @@ HAS_SERVOS              = src_filter=+<src/module/servo.cpp> +<src/gcode/control
 MORGAN_SCARA            = src_filter=+<src/gcode/scara>
 HAS_MICROSTEPS          = src_filter=+<src/gcode/control/M350_M351.cpp>
 (ESP3D_)?WIFISUPPORT    = AsyncTCP, ESP Async WebServer
-  ESP3DLib=https://github.com/luc-github/ESP3DLib.git
+  ESP3DLib=https://github.com/luc-github/ESP3DLib/archive/master.zip
   arduinoWebSockets=links2004/WebSockets@2.3.4
-  ESP32SSDP=https://github.com/luc-github/ESP32SSDP.git
+  luc-github/ESP32SSDP@^1.1.1
   lib_ignore=ESPAsyncTCP
 
 #

--- a/platformio.ini
+++ b/platformio.ini
@@ -423,7 +423,7 @@ MORGAN_SCARA            = src_filter=+<src/gcode/scara>
 HAS_MICROSTEPS          = src_filter=+<src/gcode/control/M350_M351.cpp>
 (ESP3D_)?WIFISUPPORT    = AsyncTCP, ESP Async WebServer
   ESP3DLib=https://github.com/luc-github/ESP3DLib.git
-  arduinoWebSockets=https://github.com/Links2004/arduinoWebSockets.git
+  arduinoWebSockets=https://github.com/Links2004/arduinoWebSockets/archive/2.3.4.zip
   ESP32SSDP=https://github.com/luc-github/ESP32SSDP.git
   lib_ignore=ESPAsyncTCP
 

